spec/p2p/p2p-top-level.md

```diff
@@ -0,0 +1,310 @@
+    - [Sequential Problem statement](#Sequential-Problem-statement):
+      Provides a mathematical definition of the problem statement in
+      its sequential form, that is, ignoring the distributed aspect of
+      the implementation of the blockchain.
+
+- [Part III](#part-iii---as-distributed-system): Distributed
+  aspects, system assumptions and temporal
+  logic specifications.
+
+  - [Incentives](#incentives): how faulty full nodes may benefit from
+    misbehaving and how correct full nodes benefit from cooperating.
+  
+  - [Computational Model](#Computational-Model):
+      timing and correctness assumptions.
+
+  - [Distributed Problem Statement](#Distributed-Problem-Statement):
+      temporal properties that formalize safety and liveness
+      properties in the distributed setting.
+
+- [Part IV](#part-iv---Protocol):
+  Specification of the protocols.
+
+     - [Definitions](#Definitions): Describes inputs, outputs,
+       variables used by the protocol, auxiliary functions
+
+     - [Protocol](#core-verification): gives an outline of the solution,
+       and details of the functions used (with preconditions,
+       postconditions, error conditions).
+
+     - [Liveness Scenarios](#liveness-scenarios): when the light
+       client makes progress depends heavily on the changes in the
+       validator sets of the blockchain. We discuss some typical scenarios.
+
+- [Part V](#part-v---supporting-the-ibc-relayer): Additional
+  discussions and analysis
+
+
+In this document we quite extensively use tags in order to be able to
+reference assumptions, invariants, etc. in future communication. In
+these tags we frequently use the following short forms:
+
+- TMBC: Tendermint blockchain
+- SEQ: for sequential specifications
+- LIVE: liveness
+- SAFE: safety
+- FUNC: function
+- INV: invariant
+- A: assumption
+
+--->
+
+# Part I - A Tendermint node 
+
+TODO: 
+- perhaps we should survey here what we think are the expectations of the reactors from p2p? 
+    - consensus might need for liveness that neighborhood is stable
+    - mempool might just need that a transaction can reach each validator within a reasonable amount of time (this might be achievable if at no point in time the current graph is connected, but, e.g., something along the lines that the union of the graphs over some period in time is connected.)
+- What does p2p expect from the reactors? (don't falsely report bad nodes; this puts requirements on the reactors and perhaps/likely also on the application running on top of ABCI)
+
+## Context of this document
+
+> mention other components and or specifications that are relevant for this
+spec. Possible interactions, possible use cases, etc. 
+
+> should give the reader the understanding in what environment this component
+will be used. 
+
+TODO:
+- Reactor API
+- Network? How do we communicate with other nodes
+- Discuss that validators run special set-up, and manage their own neighborhood (hide behind sentry nodes).
```

Comment:
That's a good point. Although I'd simplify to say "if you configure a node to prevent it to _normally_ connect to peers " (validator<-->sentry) then only the _normally_ connected node (the sentry) can expect the properties the overlay network provides

Reply:
Yes, I think we should use something like "normally" in the definitions of the specification. At this points I think it makes sense to talk about validators to motivate the definition

Reply:
I like the idea of defining a public network, the overlay, fully using p2p and PEX, and a private network, where each operator manually configures nodes. In fact, validators also have their IDs set as "private", so that no one learns its actual address.

Reply:
@cason says: 
1. We assume validators are somehow connected to at least a correct public node

From this point we reason about correct public nodes